README.md
# ![itty-cors](https://user-images.githubusercontent.com/865416/193479569-da9e4d63-cd0e-44da-ab00-43322bf4cd51.png)

[![Version](https://img.shields.io/npm/v/itty-cors.svg?style=flat-square)](https://npmjs.com/package/itty-cors)
[![Bundle Size](https://img.shields.io/bundlephobia/minzip/itty-cors?style=flat-square)](https://bundlephobia.com/result?p=itty-cors)
![Build Status](https://img.shields.io/github/workflow/status/kwhitley/itty-cors/build?style=flat-square)
[![Coverage Status](https://img.shields.io/coveralls/github/kwhitley/itty-cors/v0.x?style=flat-square)](https://coveralls.io/github/kwhitley/itty-cors?branch=v0.x)
[![NPM Weekly Downloads](https://img.shields.io/npm/dw/itty-cors?style=flat-square)](https://npmjs.com/package/itty-cors)
[![Open Issues](https://img.shields.io/github/issues/kwhitley/itty-cors?style=flat-square)](https://github.com/kwhitley/itty-cors/issues)

[![Discord](https://img.shields.io/discord/832353585802903572?style=flat-square)](https://discord.com/channels/832353585802903572)
[![GitHub Repo stars](https://img.shields.io/github/stars/kwhitley/itty-cors?style=social)](https://github.com/kwhitley/itty-cors)
[![Twitter](https://img.shields.io/twitter/follow/kevinrwhitley.svg?style=social&label=Follow)](https://www.twitter.com/kevinrwhitley)

Simple CORS-handling for any [itty-router](https://npmjs.com/package/itty-router) API.  Designed on Cloudflare Workers, but works anywhere.

## Features
- Tiny. Currently ~600 bytes, with zero-dependencies.
- Fully Typed/TypeScript
- Granular control over allowed methods, origins, etc.
- handles OPTIONS preflight requests
- response processor (`corsify`) can be used per-response, or globally downstream

## !! Breaking API Changes (pre v1.x)!!
You're using an early-access package.  Until itty-cors hits a stable v1.0 release, API changes will be broadcasted as minor version bumps.
- **v0.3.0** - `preflight` now **MUST** be included on the `.all` channel (all methods), rather than `.options` alone, to allow for origin-checking.
- **v0.2.0** - `allowOrigins: string` has been replaced with `origins: string[]` to accept an array of origins, and multi-origin support is now correctly implemented.
  ```js
  // previously
  const { preflight, corsify } = createCors({ allowOrigin: '*' })

  // now
  const { preflight, corsify } = createCors({ origins: ['*'] })
  ```

## Simple Usage
```js
import { Router } from 'itty-router'
import { error, json, missing } from 'itty-router-extras'
import { createCors } from 'itty-cors'

// create CORS handlers
const { preflight, corsify } = createCors()

const router = Router()

// register v2 API plus all routes
router
  .all('*', preflight)                                  // handle CORS preflight/OPTIONS requests
  .get('/version', () => json({ version: '0.1.0' }))    // GET release version
  .get('/stuff', () => json(['foo', 'bar', 'baz']))     // GET some other random stuff
  .all('*', () => missing('Are you sure about that?'))  // 404 for all else

// CF ES6 module syntax
export default {
  fetch: (...args) => router
                        .handle(...args)
                        .catch(err => error(500, err.stack))
                        .then(corsify) // cors should be applied to error responses as well
}
```

## CORS enabled on a single route (and advanced options)
```js
import { Router } from 'itty-router'
import { error, json, missing } from 'itty-router-extras'
import { createCors } from 'itty-cors'

// create CORS handlers
const { preflight, corsify } = createCors({
  methods: ['GET', 'POST', 'DELETE'], // GET is included by default... omit this if only using GET
  origins: ['*'],                     // defaults to allow all (most common).  Restrict if needed.
  maxAge: 3600,
  headers: {
    'my-custom-header': 'will be injected with each CORS-enabled response',
  },
})

const router = Router()

// register v2 API plus all routes
router
  .all('*', preflight)                                  // handle CORS preflight/OPTIONS requests
  .get('/version', () => corsify(json({ version: '0.1.0' })))  // GET release version (CORS-enabled)
  .get('/stuff', () => json(['foo', 'bar', 'baz']))     // GET some other random stuff (no CORS allowed)
  .all('*', () => missing('Are you sure about that?'))  // 404 for all else

// CF ES6 module syntax
export default {
  fetch: (...args) => router
                        .handle(...args)
                        .catch(err => error(500, err.stack))
}
```

# API

### `createCors(options?) => { preflight: function, corsify: function }`
Returns an object with two properties, `preflight` (a preflight OPTIONS middleware), and `corsify` (a response-handling function).

| Option | Type(s) | Default | Description |
| --- | --- | --- | --- |
| **origins** | `string[]` | `['*']` | By default, all origins are allowed (most common).  Modify this to restrict to specific origins.
| **maxAge** | `number` | `3600` | Set the expiry of responses
| **methods** | `string[]` | `['GET']` | Define which methods are allowed.  `OPTIONS` will be automatically added.
| **headers** | `object` | `{}` | Add any custom headers to be injected with CORS-enabled responses.

### `preflight(request: Request) => Response`
This is the preflight middleware to be injected upstream on options requests.
```js
router.options('*', preflight) // that's it!
```

### `corsify(response: Response) => Response`
This wrapper injects CORS headers into a response, if not already set (upstream).  Use this at the end of the `router.handle` Promise chain to CORS-enable all responses/the entire API, or wrap any response generator (e.g. `json()` from itty-router-extras) to make a single CORS-enabled response.


CHANGELOG.md
## Changelog
Until this library makes it to a production release of v1.x, **minor versions may contain breaking changes to the API**.  After v1.x, semantic versioning will be honored, and breaking changes will only occur under the umbrella of a major version bump.

- **v0.3.0** - `preflight` now **MUST** be included on the `.all` channel (all methods), rather than `.options` alone, to allow for origin-checking.
  ```js
  // previously
  router.options('*', preflight)

  // now
  router.all('*', preflight)
  ```
- **v0.2.0 (API CHANGE)** - `allowOrigins: string` has been replaced with `origins: string[]` to accept an array of origins, and multi-origin support is now correctly implemented.
  ```js
  // previously
  const { preflight, corsify } = createCors({ allowOrigin: '*' })

  // now
  const { preflight, corsify } = createCors({ origins: ['*'] })
  ```
- **v0.1.0** - first public release


src/itty-cors.spec.ts
import { createCors } from './itty-cors'
import { describe, expect, it } from 'vitest'

describe('fetcher', () => {
  it('exports { createCors: function } as a named export', () => {
    expect(typeof createCors).toBe('function')
  })
})


src/itty-cors.ts
interface CorsOptions {
  origins: string[],
  maxAge: number,
  methods: string[],
  headers: any,
}

export const createCors = (options?: CorsOptions) => {
  const {
    origins = ['*'],
    maxAge = 3600,
    methods = ['GET'],
    headers = {},
  } = options

  let allowOrigin

  const responseHeaders = {
    'content-type': 'application/json',
    'Access-Control-Allow-Methods': methods.join(', '),
    'Access-Control-Max-Age': maxAge,
    ...headers,
  }

  const preflight = (r) => {
    const useMethods = [...new Set(['OPTIONS', ...methods])]

    const origin = r.headers.get('origin')

    // set allowOrigin globally
    allowOrigin = (origins.includes(origin) || origins.includes('*')) &&
      { 'Access-Control-Allow-Origin': origin }

    if (r.method === 'OPTIONS') {
      // Make sure the necessary headers are present
      // for this to be a valid pre-flight request
      if (
        r.headers.get('Origin') !== null &&
        r.headers.get('Access-Control-Request-Method') !== null &&
        r.headers.get('Access-Control-Request-Headers') !== null
      ) {
        // Handle CORS pre-flight request.
        // If you want to check or reject the requested method + headers
        // you can do that here.
        const headers = {
          ...responseHeaders,
          'Access-Control-Allow-Methods': useMethods.join(', '),
          'Access-Control-Allow-Headers': r.headers.get('Access-Control-Request-Headers'),
          ...allowOrigin,
        }

        return new Response(null, { headers })
      }

      // Handle standard OPTIONS request.
      // If you want to allow other HTTP Methods, you can do that here.
      return new Response(null, {
        headers: {
          Allow: useMethods.join(', '),
        },
      })
    }
  }

  const corsify = (response) => {
    const { headers, status, body } = response

    if (headers.get('access-control-allow-origin') || [301, 302, 308].includes(status)) {
      return response // terminate immediately if CORS already set
    }

    return new Response(body, {
      status,
      headers: {
        ...headers,
        ...responseHeaders,
        ...allowOrigin,
      },
    })
  }

  return {
    corsify,
    preflight,
  }
}


index.js
import { Router } from 'itty-router'
import { error, json, missing, status } from 'itty-router-extras'

// response-modifying middleware to add CORS headers (including in OPTIONS requests)
const addCorsheaders = response => {
  response.headers.set('Access-Control-Allow-Origin', '*')
  response.headers.set('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS')
  response.headers.set('Access-Control-Allow-Headers', 'authorization, referer, origin, content-type')
  response.headers.set('Access-Control-Max-Age', '3600')

  return response
}

// create a router
const router = Router()

// 204 CORS response for all options requests
router.options('*', () => status(204))

// add routes
router.get('/test', () => json({ success: true }))

// 404 for all other routes
router.get('*', () => missing('Are you sure about that?'))

export default {
  fetch: (...args) => router
                        .handle(...args)
                        .then(addCorsheaders)
                        .catch(error),
}
